backend/remote-state/artifactory/client.go

```diff
@@ -48,16 +79,88 @@ func (c *ArtifactoryClient) Get() (*remote.Payload, error) {
 }
 
 func (c *ArtifactoryClient) Put(data []byte) error {
+	var err error
+	logStart("Put")
+	defer logResult("Put", &err)
+
 	p := fmt.Sprintf("%s/%s/%s", c.repo, c.subpath, ARTIF_TFSTATE_NAME)
-	if _, err := c.nativeClient.Put(p, string(data), make(map[string]string)); err == nil {
+	if _, err = c.nativeClient.Put(p, string(data), make(map[string]string)); err == nil {
 		return nil
 	} else {
 		return fmt.Errorf("Failed to upload state: %v", err)
 	}
 }
 
 func (c *ArtifactoryClient) Delete() error {
+	var err error
+	logStart("Delete")
+	defer logResult("Delete", &err)
 	p := fmt.Sprintf("%s/%s/%s", c.repo, c.subpath, ARTIF_TFSTATE_NAME)
-	err := c.nativeClient.Delete(p)
+	err = c.nativeClient.Delete(p)
+	return err
+}
+
+func (c *ArtifactoryClient) Lock(info *state.LockInfo) (string, error) {
+	if c.lockUrl == "" {
+		return "", nil
+	}
+	c.lockID = ""
+	var err error
+	var output []byte
+	logStart("Lock")
+	defer logResult("Lock", &err)
+	rand.Seed(time.Now().UnixNano())
+	info.ID = fmt.Sprintf("%s-%d", info.ID, rand.Intn(100000000))
+	jsonLockInfo := info.Marshal()
+	p := fmt.Sprintf("%s/%s/%s", c.lockRepo, c.lockSubpath, ARTIF_TFLOCK_NAME)
+	if _, err = c.lockNativeClient.Put(p, string(jsonLockInfo), make(map[string]string)); err == nil {
+		if c.lockReadbackWait <= 0 {
+			c.lockID = info.ID
+			c.jsonLockInfo = jsonLockInfo
+			return info.ID, nil
+		}
+		time.Sleep(time.Duration(c.lockReadbackWait) * time.Millisecond)
+		// readback and compare with original info
+		output, err = c.lockNativeClient.Get(p, make(map[string]string))
+		if err != nil {
+			return "", fmt.Errorf("Failed to read back lockInfo. Failed to lock: %v", err)
+		} else {
+			var lockInfoReadBack state.LockInfo
+			err = json.Unmarshal(output, &lockInfoReadBack)
+			if err != nil {
+				return "", fmt.Errorf("Failed to Unmarchal lockInfo from readback. Failed to lock: %v", err)
+			} else {
+				if reflect.DeepEqual(*info, lockInfoReadBack) {
+					c.lockID = info.ID
+					c.jsonLockInfo = jsonLockInfo
+					return info.ID, nil
+				} else {
+					err = errors.New("lockinfo readback and original are not the same. Failed to lock")
+					return "", err
+				}
+			}
+		}
+	} else {
+		return "", fmt.Errorf("Failed to lock: %v", err)
+	}
+}
+
+func (c *ArtifactoryClient) Unlock(id string) error {
+	if c.lockUrl == "" {
```

Comment:
The current contract requires that an incorrect ID fail to unlock the backend. It doesn't appear to be checking the ID value here.